docs/architecture/anti-entropy/hinted-handoff.rst

```diff
@@ -43,4 +43,34 @@ Hinted handoff is enabled and managed by these settings in :code:`scylla.yaml`:
 
+Sync point HTTP API
+^^^^^^^^^^^^^^^^^^^
+
+ScyllaDB provides you with a means to wait until hints for given nodes have been replayed. It's realized by creating a so-called sync point that corresponds to the hints that have accumulated up to the time of its creation. It's always created for both regular hints and updates for materialized views. You can create and await sync points by sending HTTP requests to :code:`http://{API_ADDRESS}:{API_PORT}/hinted_handoff/sync_point/`, where :code:`{API_ADDRESS}` corresponds to the API IP address of the coordinator node which hints are supposed to be sent from.
+
+You can create a sync point by sending a POST request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`target_hosts`: a list of the IP addresses of nodes towards which hints should be replayed; the IPs are divided by commas, e.g. :code:`127.0.0.2,127.0.0.3`. If no IP address is specified, ScyllaDB assumes the sync point should be created towards all live nodes in the cluster. This parameter is not required -- if it's not present, ScyllaDB will treat the request as if zero IP addresses were specified.
+
+ScyllaDB should return a JSON containing the ID of the created sync point as the response.
+
+You can await a previously created sync point by sending a GET request to :code:`/hinted_handoff/sync_point` with the following parameters:
+
+* :code:`id`: the ID of the sync point the user wants to await. This parameter is mandatory. Only one ID can be passed,
+* :code:`timeout`: an integer value specifying how many seconds ScyllaDB should wait for the hints corresponding to the sync point to be sent. If the value is negative, ScyllaDB will wait indefinitely long. This parameter is not mandatory; if it's not present, ScyllaDB will return the result immediately.
+
+The database should return a JSON containing one of the two values as the response:
+
+* :code:`"DONE"` -- if all of the hints corresponding to the passed sync point have already been sent,
+* :code:`"IN_PROGRESS"` -- if some of those hints have not been sent yet.
+
+In case of an invalid request or an error when attempting to create or await a sync point, a JSON reporting the issue will be returned as the response.
+
+.. note:: If the node is turned off while you're awaiting a sync point on it, you'll receive an error JSON message as the response. You should try awaiting the sync point again when the node has been turned on again.
+
+Sync points aren't a fully reliable mechanism. If you receive :code:`"IN_PROGRESS"` as the response, you have a guarantee that not all of the hints corresponding to the passed sync point have been sent yet. However, if you receive :code:`"DONE"`, it may happen that **not** all of the nodes specified by :code:`target_hosts` when creating the sync point have received all of the hints.
```

Comment:
Yes. Consider the following scenario:

1. Start node A with hinted handoff enabled for all DCs.
2. Save hints on node A for nodes B and C, where B belongs to the same DC as A – DC1, but C belongs to another one – DC2.
3. Create a sync point on node A for nodes B and C.
4. Turn off node A.
5. Start node A with hinted handoff enabled only for DC1. Because of that, only an endpoint manager corresponding to node B will be created.
6. Nodes B and C become reachable from node A again. Since there's an endpoint manager only for node B, only hints targeting that node will be sent.
7. The user awaits the previously created sync point and Scylla reports they've all been sent because for all existing endpoint managers, we've sent all of the hints saved until the moment when the sync point was created.

The issue is that we don't verify that there's an endpoint manager for every IP a sync point corresponds to, so we might end up checking only a subset of those.

I didn't give this much thought, but I think there might be a solution.

**Solution 1.** At the beginning of `manager::wait_for_sync_point()`, verify that we've created endpoint managers for all of the endpoints that the given sync point is concerned about. If not all of them exist, scan the disk and create the missing ones.

Potential problems:

1. We'd need to analyse if this would interact well during the transition period from IP-based to host-ID-based hinted handoff. Chances are this solutions might be difficult to implement in those circumstances because of the mappings IP–host ID we keep track of. For example, what if there's no mapping IP–host ID in the token metadata for a given IP? Should we ignore such a hint directory? Or should we not?

2. It could break the invariant that we only send hints to the DCs specified by the currently used host filter. This shouldn't be that big of an issue, i.e. I guess it would be fairly easy to implement, but it's definitely something to keep in mind.

**Solution 2.** Always create endpoint managers for all existing hint directories.

Potential problems: pretty much the same as above, except this strategy would proboably be easier to implement.

I think it would be best to leave the current implementation as-is until we get to 6.2 and can drop support for IP-based hinted handoff. We'd invalidate all of the created sync points before that version and change the parameter `target_hosts` to host IDs. Then solving the problems should be at least a bit easier.

When thinking about changing the implementation, one other scenario worth keeping in mind is this:

1. Start node A.
2. Save hints on node A for nodes B and C.
3. Create a sync point on node A for B and C.
4. B is, for some reason, drained on A and all of its hints are sent. The hint directory is deleted as well.
5. Restart A. An endpoint manager is created for C, but not for B (because there is no hint directory corresponding to it anymore).
6. C becomes reachable from A again and we send all of the hints to it.
7. The user awaits the sync point.

Currently, we'd (correctly) return `"DONE"` here.

Reply:
Please create an issue and post this description there. This PR's review is not the right place to discuss the details of the bug. The most important thing is that it's not a design limitation, it looks like this bug can be fixed.

Every feature has, or can have, bugs. We might encounter a bug where `IN_PROGRESS` keeps being returned even after all hints were sent. Are you willing to update the documentation every time new bugs appear?

(Btw, this property might not be considered true right now - imagine that it takes 100ms to send this value back to the client, but 50ms after the response was generated hints were replayed - I don't consider that a bug)

Reply:
I see, that makes sense. I'll remove this fragment then.

Reply:
Issue created: #19703